### PR TITLE
[enh] Add SpatialTransform.points() and PointSetTransformer module

### LINKED ISSUE #53: Add PointTransformer which maps points from any coordinate system to another
Similar to an `ImageTransformer`, a `PointTransformer` could decorate an optional `SpatialTransform` and map points from any coordinate system specified during its construction to another or the same coordinate system. For example, from world to world coordinate, or fixed target image grid indices to moving image grid indices. When no `SpatialTransform` is given, the identity is assumed and the `PointTransform` performs the same operation as `deepali.core.grid.grid_transform_points()`.

CC @qiuhuaqi

### PR DESCRIPTION
**Pull request by aschuh-hf**
The `SpatialTransform.points()` method can be used instead of `SpatialTransform.forward()` when the input point coordinates are not normalized with respect to the `SpatialTransform.grid()` and `SpatialTransform.axes()`, or when the output point coordinates should be mapped to a different domain as part of applying this function. This is mainly a convenience function, but should also make applying spatial transformations to unnormalized points (e.g., w.r.t. to the world coordinate system) more clear.

In addition, the `PointSetTransformer` module calls the `SpatialTransform.points()` function to transform a given input. The arguments of the `SpatialTransform.points()` method are hereby set during transformer module initialization.

Closes #53.

### CODE REVIEW

**Comment by qiuhuaqi** [comment]
@aschuh-hf 
I noticed that this `SpatialTransform.points()` function maps the input points first to `self.grid()` in 
```
points = grid.transform_points(
            points,
            axes=axes,
            to_grid=self.grid(),
            to_axes=self.axes(),
            decimals=None,
        )
```
then apply the transformation in the normalized space using `points = self.forward(points)`. 

Just by looking at the input signature, it seems like the potentially un-normalized input points are being mapped onto the grid of `self.grid()` which comes with the `SpatialTransform`. But this grid might not be "normalized" grid that's required by `self.forward()`. For example, sometimes this grid is the same as a target image grid. This would mean a mismatch between the space/domain expected by the `forward()` function and the points. 

Am I missing something here? Or perhaps the `transform_points(...)` step actually maps to a normalized space?

**Comment by aschuh-hf** [comment]
Not every `transform` passed to `SpatialTransformer` must work in normalized space. When used in classic test-time optimization registration, it might be transforming from world to world.

The source and target `Grid` define the domain and mapping between coordinates (world, image, cube), but the `axes` and `to_axes` determine which of these coordinates are passed as input to the `transform` and expected as output of the `SpatialTransformer`, respectively. The default is for `axes` and `to_axes` to mean cube coordinates, and which specifically is determined by the `align_corners` attribute:

https://github.com/BioMedIA/deepali/blob/03e8cbb08c90738c8a14a1feb2959be8cbfef78c/src/deepali/spatial/base.py#L108-L115

The forward method of the `PointSetTransformer` maps the input coordinates to the `axes` expected by the `transform`, and then maps the transformed coordinates to the requested `to_axes` representation.

https://github.com/BioMedIA/deepali/blob/03e8cbb08c90738c8a14a1feb2959be8cbfef78c/src/deepali/spatial/transformer.py#L283-L301

**Comment by qiuhuaqi** [comment]
Thanks for the speedy reply!  🚀

I understand that `SpatialTransformer` handles the conversion between the axes. But when `SpatialTransform` is used by itself, things feel less clear to me.

For example, if the `transform` maps between world coordinates (which is my usecase now), then isn't there an inherit mismatch with the `transform.axes()`, unless we allow the user to set this to the world axes? 

Also, if the `transform.points()` of the same transform is called with world coordinates as input, then the points are first mapped to `self.axes()` which is set to cube coordinates as you mentioned (with no setter to change this) and wouldn't be correctly transformed by `transform.forward()` which now operates in world coordinates. 

https://github.com/BioMedIA/deepali/blob/03e8cbb08c90738c8a14a1feb2959be8cbfef78c/src/deepali/spatial/base.py#L277-L283


I feel like it might be better to allow users to set `SpatialTransform.axes` somehow so `.forward()` and `.points()` are consistent. Or still enforce cube coordinates within `SpatialTransform` but simply only expose `SpatialTransform.points()` as forward method instead. 

This seems to be related to: https://github.com/BioMedIA/deepali/issues/52

**Comment by aschuh-hf** [comment]
You're right, this is definitely continuing on #52.

> Or still enforce cube coordinates within SpatialTransform but simply only expose SpatialTransform.points() as forward method instead.

In your use case, could you use `PointSetTransformer` as this transform wrapper which effectively makes `SpatialTransform.points()` the forward method of the `.nn.Module`? Or do you need a module whose type is part of the `SpatialTransform` class hierarchy?

**Comment by qiuhuaqi** [comment]
Indeed this wrapping should work. 👍 
I'm customising a workflow based on the FFD example, re-using the `example.ffd.losses.PairwiseImageRegistrationLoss` class. While combining image-based losses with point-based losses, I need to first map the original target and source points to target grid and cube coordinates if I don't change the `self.transform(points)` calls. 

https://github.com/BioMedIA/deepali/blob/03e8cbb08c90738c8a14a1feb2959be8cbfef78c/examples/ffd/losses.py#L274-L291

I think there is no problem doing this for now. We can continue the discussion in #52

**Review by qiuhuaqi — APPROVED** [review]

**Comment by aschuh-hf** [comment]
Another workaround would be to create your own SpatialTransform subclass which wraps another just like PointSetTransformer does. Your subclass then just calls `self.transform.points()` in its `forward()` method.

But considering you are going to optimize the transform parameters, I wonder if it isn't best to just once map your input landmarks / point sets to normalized coordinates, optimize a transform with respect to these, and then map the final optimized transform parameters to a transformation of world coordinates. That way you avoid all the back-and-forth coordinate conversions during the optimization. Is there a benefit to keep the input points in world coordinates?

**Comment by qiuhuaqi** [comment]
Yes I did exactly that mapping and optimised the transformation. I used the grid methods to map all landmarks to the same coordinate and target space, and everything worked well. With `target_landmarks` and `source_landmarks` originally in `Axes.GRID` space with their respective grids: 

```python
assert target.grid().axes() in [Axes.CUBE, Axes.CUBE_CORNERS]
target_landmarks_norm = target.grid().transform_points(
    target_landmarks, 
    axes=Axes.GRID, 
    to_axes=target.grid().axes(),  # to_grid=None
)
source_landmarks_norm = source.grid().transform_points(
    source_landmarks, 
    axes=Axes.GRID, 
    to_axes=target.grid().axes(), 
    to_grid=target.grid()
)
```

The only thing is that the order of the original coordinates (with `Axes.GRID`) need to be in z-y-x order (SimpleITK convention?), which caught me a little by surpise. Because I extracted these coordinates form volumetric labels so they are originally in array index. Flipping the spatial dimension worked. 

I don't know if this is just a convention that I was not aware of. Maybe we can update the documentation for `Axes.GRID` to make this clearer? 
https://biomedia.github.io/deepali/reference/core/domain.html
